File: SSUMobile/Modules/Maps/SSUMapModule.m
//
//  SSUMapModule.m
//  SSUMobile
//
//  Created by Eric Amorde on 9/8/15.
//  Copyright (c) 2015 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUMapModule.h"
#import "SSULogging.h"
#import "SSUPointsBuilder.h"
#import "SSUBuildingPerimetersBuilder.h"
#import "SSUConnectionsBuilder.h"
#import "SSUMoonlightCommunicator.h"
#import "SSUConfiguration.h"

@implementation SSUMapModule

+ (instancetype) sharedInstance {
    static SSUMapModule * instance = nil;
    static dispatch_once_t token;
    dispatch_once(&token, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

#pragma mark - SSUModule

- (nonnull NSString *) title {
    return NSLocalizedString(@"Map",
                             @"The campus Map shows the location of campus buildings and provides directions");
}

- (nonnull NSString *) identifier {
    return @"campusmap";
}

- (UIView *) viewForHomeScreen {
    return [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"map_icon"]];
}

- (UIImage *) imageForHomeScreen {
    return [UIImage imageNamed:@"map_icon"];
}

- (UIViewController *) initialViewController {
    UIStoryboard * storyboard = [UIStoryboard storyboardWithName:@"Map_iPhone"
                                                          bundle:[NSBundle bundleForClass:[self class]]];
    return [storyboard instantiateInitialViewController];
}

- (BOOL) showModuleInNavigationBar {
    return NO;
}

- (BOOL) shouldNavigateToModule {
    return YES;
}

- (void) setup {
    NSManagedObjectModel * model = [self modelWithName:@"Map"];
    NSPersistentStoreCoordinator * coordinator = [self persistentStoreCoordinatorWithName:@"Map" model:model];
    self.context = [self contextWithPersistentStoreCoordinator:coordinator];
    self.backgroundContext = [self backgroundContextFromContext:self.context];
}

- (void) clearCachedData {
    //TODO: implement clearCachedData
}

- (void) updateData:(void (^)())completion {
    SSULogDebug(@"Update Map NEW");
    __block int count = 0;
    int total = 3;
    [self updatePoints:^{
        count++;
        if (count == total && completion) {
            completion();
        }
    }];
    [self updatePerimeters:^{
        count++;
        if (count == total && completion) {
            completion();
        }
    }];
    [self updateConnections:^{
        count++;
        if (count == total && completion) {
            completion();
        }
    }];
}

- (void) updatePoints:(void (^)())completion {
    if (completion) completion();
    return;
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUMapPointsUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"points" sinceDate:date completion:^(id json, NSError *error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update Map points: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [self.backgroundContext performBlock:^{
                [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUMapPointsUpdatedDateKey];
                [self buildPointsJSON:json];
                if (completion) {
                    completion();
                }
            }];
        }
        SSULogDebug(@"Finish %@",self.title);
    }];
}

- (void) updatePerimeters:(void (^)())completion {
    if (completion) completion();
    return;
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUMapPerimetersUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"perimeters" sinceDate:date completion:^(id json, NSError *error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update Map perimeters: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [self.backgroundContext performBlock:^{
                [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUMapPerimetersUpdatedDateKey];
                [self buildPerimetersJSON:json];
                if (completion) {
                    completion();
                }
            }];
        }
    }];
}

- (void) updateConnections:(void (^)())completion {
    if (completion) completion();
    return;
    NSDate * date = [[SSUConfiguration sharedInstance] dateForKey:SSUMapPerimetersUpdatedDateKey];
    [SSUMoonlightCommunicator getJSONFromPath:@"connections" sinceDate:date completion:^(id json, NSError *error) {
        if (error != nil) {
            SSULogError(@"Error while attemping to update Map connections: %@", error);
            if (completion) {
                completion();
            }
        }
        else {
            [self.backgroundContext performBlock:^{
                [[SSUConfiguration sharedInstance] setDate:[NSDate date] forKey:SSUMapPerimetersUpdatedDateKey];
                [self buildConnectionsJSON:json];
                if (completion) {
                    completion();
                }
            }];
        }
    }];
}

#pragma mark - Private

- (void) buildPointsJSON:(id)json {
    SSUPointsBuilder * builder = [[SSUPointsBuilder alloc] init];
    builder.context = [self backgroundContextFromContext:self.context];
    [builder build:json];
}

- (void) buildPerimetersJSON:(id)json {
    SSUBuildingPerimetersBuilder * builder = [[SSUBuildingPerimetersBuilder alloc] init];
    builder.context = self.backgroundContext;
    [builder build:json];
}

- (void) buildConnectionsJSON:(id)json {
    SSUConnectionsBuilder * builder = [[SSUConnectionsBuilder alloc] init];
    builder.context = self.backgroundContext;
    [builder build:json];
}

@end

File: SSUMobile/Modules/Maps/SSUMapModule.swift
//
//  SSUMapModule.swift
//  SSUMobile
//
//  Created by Eric Amorde on 4/9/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation

class SSUMapModule: SSUCoreDataModuleBase, SSUModuleUI {
    
    @objc(sharedInstance)
    static let instance = SSUMapModule()
    
    // MARK: SSUModule
    
    var title: String {
        return NSLocalizedString("Map", comment: "The campus Map shows the location of campus buildings and provides directions")
    }
    
    var identifier: String {
        return "campusmap"
    }
    
    func setup() {
        setupCoreData(modelName: "Map", storeName: "Map")
    }
    
    func updateData(_ completion: (() -> Void)? = nil) {
        SSULogging.logDebug("Update Map")
        updatePoints {
            self.updatePerimeters {
                self.updateConnections {
                    completion?()
                }
            }
        }
    }
    
    func updatePoints(completion: (() -> Void)? = nil) {
        let date = SSUConfiguration.sharedInstance().date(forKey:SSUMapPointsUpdatedDateKey)
        SSUMoonlightCommunicator.getJSONFromPath("ssumobile/map/point/", since:date) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while attemping to update Map points: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUMapPointsUpdatedDateKey)
                self.buildPoints(json) {
                    completion?()
                }
            }
        }
    }
    
    func updatePerimeters(completion: (() -> Void)? = nil) {
        SSUMoonlightCommunicator.getJSONFromPath("ssumobile/map/perimeter/", since:nil) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while attemping to update Map perimeters: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUMapPerimetersUpdatedDateKey)
                self.buildPerimeters(json) {
                    completion?()
                }
            }
        }
    }
    
    func updateConnections(completion: (() -> Void)? = nil) {
        SSUMoonlightCommunicator.getJSONFromPath("ssumobile/map/point_connection/", since:nil) { (response, json, error) in
            if let error = error {
                SSULogging.logError("Error while attemping to update Map connections: \(error)")
                completion?()
            } else {
                SSUConfiguration.sharedInstance().setDate(Date(), forKey: SSUMapPerimetersUpdatedDateKey)
                self.buildConnections(json) {
                    completion?()
                }
            }
        }
    }

    
    func buildPoints(_ json: Any, completion: (() -> Void)? = nil) {
        let builder = SSUPointsBuilder()
        builder.context = backgroundContext
        builder.context.perform {
            builder.build(json)
            DispatchQueue.main.async {
                completion?()
            }
        }
    }
    
    func buildPerimeters(_ json: Any, completion: (() -> Void)? = nil) {
        let builder = SSUBuildingPerimetersBuilder()
        builder.context = backgroundContext
        builder.context.perform {
            builder.build(json)
            DispatchQueue.main.async {
                completion?()
            }
        }
    }
    
    func buildConnections(_ json: Any, completion: (() -> Void)? = nil) {
        let builder = SSUConnectionsBuilder()
        builder.context = backgroundContext
        builder.context.perform {
            builder.build(json)
            DispatchQueue.main.async {
                completion?()
            }
        }
    }
    
    // MARK: SSUModuleUI
    
    func imageForHomeScreen() -> UIImage? {
        return UIImage(named: "map_icon")
    }
    
    func viewForHomeScreen() -> UIView? {
        return nil
    }
    
    func initialViewController() -> UIViewController {
        let storyboard = UIStoryboard(name: "Map_iPhone", bundle: Bundle(for: type(of: self)))
        return storyboard.instantiateInitialViewController()!
    }
    
    func shouldNavigateToModule() -> Bool {
        return true
    }
    
    func showModuleInNavigationBar() -> Bool {
        return false
    }
}
